packages/key-did-resolver/src/__tests__/secp256r1.test.ts

```diff
@@ -0,0 +1,19 @@
+import varint from "varint"
+import multibase from "multibase"
+import * as mapper from "../secp256r1"
+
+describe('Secp256r1 mapper', () => {
+
+    it('successfully resolves the document from did', async () => {
+        const id = "zruuPojWkzGPb8sVc42f2YxcTXKUTpAUbdrzVovaTBmGGNyK6cGFaA4Kp7SSLKecrxYz8Sc9d77Rss7rayYt1oFCaNJ"
```

Comment:
Afaik the Key DIDs should only contain the X value (see secp256k1 and ed25519 examples).

Reply:
does the transmute-industries/did-key.js also use this format?

Reply:
If so this should be alright.

Reply:
Transmute industries may be inconsistent with the present state. I believe Orie is updating a part before becoming consistent with the did method key spec and this issue.
https://github.com/w3c-ccg/did-method-key/issues/29#issuecomment-786028270

However, the did key is the same length as in: https://w3c-ccg.github.io/did-method-key/#p-256 .

Reply:
edit: Orie is waiting on bls12381 which is not working with node 15. https://github.com/transmute-industries/did-key.js/pull/78#issuecomment-806200291

Reply:
Fwiw, I am going back to see how I created the did:key. The code works with the did:key examples from the ccg spec and the ones that I created on my own.

Reply:
This is my code to create the did:key :: https://github.com/bshambaugh/key-did-provider-secp256r1

Reply:
Yes. The hex that I used to create the key is 64 bytes. It is a hex string that has a uint8array of bytes that looks like this: https://www.youtube.com/watch?v=_MjjF211BM8&t=263s .

Looking at the code and output suggests that 64 bytes is the full X,Y key. 
https://gist.github.com/bshambaugh/36aa03e0d79b630678c558f32c6ed936

Reply:
I can ask if the compressed public key would be better for did:key

Reply:
Yeah, I'm mostly just surprised why `secp256k1` is compressed while `r1` is not. I'm fine with leaving it as is though since it seems like that's what the did:key spec says.

Reply:
Kyle Hartog believed the spec was the best thing to look at. He also believed that the work was stalled. 
https://lists.w3.org/Archives/Public/public-credentials/2021Apr/0171.html
I am currently investigating how difficult it would be to modify the function to convert the compressed form into the uncompressed form for resolving the did doc.

Reply:
There is some conversation going on over here about compressed vs. uncompressed did:keys. I am linking to it to keep track of it all: https://github.com/w3c-ccg/did-method-key/issues/32

Reply:
Thanks. Should we wait for that to be resolved before we merge this?

Reply:
The last link [1] that dlongley linked to in his Post is interesting.

[1] https://stackoverflow.com/questions/17171542/algorithm-for-elliptic-curve-point-compression/30431547#30431547
[2] https://github.com/w3c-ccg/did-method-key/issues/32#issuecomment-827681999

I found out that modPow is used for the decompression. Presently I am trying to perform the same functions by not using the JavaScript Big Integer Library. modPow is for modular exponentiation [4].

[3] https://github.com/peterolson/BigInteger.js
[4] https://en.wikipedia.org/wiki/Modular_exponentiation

BigInt [5] is included natively in JavaScript now, which allows me to emulate some of the functionality of the [1] link.

However, I am struggling to implement modPow. Assuming:

var yBig = xBig.pow(3).sub( xBig.multiply(3) ).add( b ).modPow( pIdent, prime );

is the same as

var yBig = ((xBig**3n - xBig*3n + b)**pIdent) % prime;

gives me an error: "RangeError: Maximum BigInt size exceeded"

To solve this error, I looked for implementations of modPow with JavaScript. I found this one [6] that works with BigInt.

[5] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
[6] https://github.com/juanelas/bigint-mod-arith

or something like this [7] if the library is to be avoided

[7] https://gist.github.com/krzkaczor/0bdba0ee9555659ae5fe

Should I include a library, or should I add a function for implementing BigInt to my code?

Reply:
> Should I include a library, or should I add a function for implementing BigInt to my code?

Which library are you referring to? If it's not too big then I think it's fine to include.

Reply:
I have raw (64 byte), compressed (33 byte), and uncompressed (65 byte) keys in my code now. My tests pass, but I do not have full coverage yet.
![scr-1620860401](https://user-images.githubusercontent.com/4206177/118070149-fc8df880-b36a-11eb-9fa4-829df22c7702.png)

Reply:
Ahh...line 61 was for the function ECPointDecompress . I added an extra test and it looks to be covered. The conditional if block that line 61 is in only executes if the parity of the compressed key is opposite (of what is should be?).

Reply:
![scr-1621014068](https://user-images.githubusercontent.com/4206177/118312749-6e735880-b4b7-11eb-9f20-e889a3d3b770.png)

Now I need to address the pubKeyBytesToXY function. I believe the function will fail if the bytesCount is 65 and the 1st byte is not 03, the bytes count is 33 and the 1st bytes is not 03 or 02, or the bytes count is not 64, 65, or 33. I guess I could have the function return null if this is the case. I am not sure how the did:key document will resolve if the function returns null. I could try to report null in place of the key, or give an error message instead of resolving the document will nulls.

Reply:
![scr-1621027899](https://user-images.githubusercontent.com/4206177/118333903-2dd50880-b4d2-11eb-8d7f-11e73412e21f.png)

I have full coverage. Unfortunately, I found some tests that I can create that can fail. I have not included them yet since they use functions included to enable the tests, not functions that are in the code. I think that the same bug may be in the code too. I am going to try to see if I can fix the bug in the test function and the code functions.

test that fail:

test('test a hex string shorter than 33 bytes', () => {
   const inputPublicKeyHex = '36f8964623378bdc068d4bce07ed17c8fa486f9ac0c2613ca3c8c306d7bb6'
   const publicKey_u8a = pubKeyHexToUint8Array(inputPublicKeyHex);
   const pubKeyBytesToXY = mapper.pubKeyBytesToXY(publicKey_u8a);
   expect(pubKeyBytesToXY).toEqual(null);
});


test('test a hex string between 64 and 33 bytes', () => {
   const inputPublicKeyHex = '36f8964623378bdc068d4bce07ed17c8fa486f9ac0c2613ca3c8c306d7bb61cd36717b8ac5e4fea8ad23dc8d0783c2318ee4ad7a80db6e0026ad0b072a24f'
   const publicKey_u8a = pubKeyHexToUint8Array(inputPublicKeyHex);
   const pubKeyBytesToXY = mapper.pubKeyBytesToXY(publicKey_u8a);
   expect(pubKeyBytesToXY).toEqual(null);
});